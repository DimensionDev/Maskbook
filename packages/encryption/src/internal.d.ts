type Readwrite<T> = {
    -readonly [key in keyof T]: T[key]
}
