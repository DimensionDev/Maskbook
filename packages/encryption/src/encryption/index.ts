export * from './Decryption'
export * from './Encryption'
export * from './AppendEncryption'
