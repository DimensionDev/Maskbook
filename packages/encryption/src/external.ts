// This file is the entry for external usage.
export {
    encodePayload,
    parsePayload,
    EC_KeyCurveEnum,
    SocialNetworkEnum,
    SocialNetworkEnumToProfileDomain,
    type EC_Key,
    type Signature,
    type PayloadParseResult,
    type PayloadWellFormed,
    type SupportedPayloadVersions,
} from './payload/index.js'

export {
    decrypt,
    DecryptError,
    DecryptProgressKind,
    DecryptIntermediateProgressKind,
    DecryptErrorReasons,
    type DecryptOptions,
    type DecryptIO,
    type DecryptEphemeralECDH_PostKey,
    type DecryptStaticECDH_PostKey,
    type DecryptProgress,
    type DecryptIntermediateProgress,
    type DecryptReportedInfo,
    type DecryptSuccess,
} from './encryption/index.js'

export { socialNetworkEncoder, socialNetworkDecoder } from './social-network-encode-decode/index.js'

export { type TypedMessage, type Meta } from '../../typed-message/dist/base/index.js'
export * from '../../typed-message/dist/base/core/index.js'
export {
    ProfileIdentifier,
    PostIVIdentifier,
    CheckedError,
    OptionalResult,
    type AESCryptoKey,
    type EC_Public_CryptoKey,
    type EC_Private_CryptoKey,
    type EC_CryptoKey,
} from '@masknet/shared-base'

export * from 'ts-results-es'
