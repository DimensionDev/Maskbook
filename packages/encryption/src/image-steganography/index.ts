import { encode, decode, GrayscaleAlgorithm } from '@dimensiondev/stego-js'
import type { EncodeOptions } from '@dimensiondev/stego-js'
import { getDimension } from './utils.js'
import { getPreset, findPreset } from './presets.js'
import { encodeArrayBuffer } from '@dimensiondev/kit'

export { GrayscaleAlgorithm, AlgorithmVersion } from '@dimensiondev/stego-js'

export interface SteganographyIO {
    downloadImage: (url: string) => Promise<ArrayBuffer>
}
export interface EncodeImageOptions extends SteganographyIO {
    data: string | Uint8Array
    password: string
    grayscaleAlgorithm?: GrayscaleAlgorithm
    preset: SteganographyPreset
}
export enum SteganographyPreset {
    Preset2021 = '2021',
    Preset2022 = '2022',
}

export async function steganographyEncodeImage(buf: ArrayBuffer, options: EncodeImageOptions) {
    const { downloadImage, data, password, grayscaleAlgorithm } = options
    const preset = getPreset(options.preset)
    if (!preset) throw new Error('Failed to find preset.')

    const optionalOptions: Partial<Readwrite<EncodeOptions>> = {}
    if (grayscaleAlgorithm) optionalOptions.grayscaleAlgorithm = grayscaleAlgorithm

    const text = typeof data === 'string' ? data : encodeArrayBuffer(data)

    return new Uint8Array(
        await encode(buf, await downloadImage(preset.mask), {
            ...preset.options,
            ...optionalOptions,
            text,
            pass: password,
        }),
    )
}

export type DecodeImageOptions = SteganographyIO & Partial<EncodeOptions> & Pick<EncodeOptions, 'pass'>

export async function steganographyDecodeImage(image: Blob | string, options: DecodeImageOptions) {
    const buffer = typeof image === 'string' ? await options.downloadImage(image) : await image.arrayBuffer()
    const dimension = getDimension(buffer)
    const preset = findPreset(dimension)
    if (!preset) return ''
    return decode(buffer, await options.downloadImage(preset.mask), {
        ...preset.options,
        ...options,
    })
}
