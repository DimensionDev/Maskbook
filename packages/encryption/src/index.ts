export { AESAlgorithmEnum, PublicKeyAlgorithmEnum, parsePayload, encodePayload } from './payload'
export type { AESKey, AsymmetryCryptoKey, PayloadParseResult, PayloadWellFormed, Signature } from './payload'
