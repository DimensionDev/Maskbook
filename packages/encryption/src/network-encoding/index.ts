import { EncryptPayloadNetwork } from '../payload/types.js'
import { sharedDecoder, sharedEncoder } from './shared.js'
import { TwitterDecoder, __TwitterEncoder } from './twitter.js'

export function decodeByNetwork(network: EncryptPayloadNetwork, content: string): Array<string | Uint8Array> {
    if (network === EncryptPayloadNetwork.Twitter) {
        return TwitterDecoder(content)
            .map((x) => [x])
            .unwrapOr([])
    }
    const possiblePayload = content.match(/(\u{1F3BC}[\w+/=|]+:\|\|)/giu) || []

    const result: Array<string | Uint8Array> = []
    for (const payload of possiblePayload) {
        const decoded = sharedDecoder(payload)
        if (decoded.isSome()) result.push(decoded.value)
        else result.push(payload)
    }
    return result
}
export function encodeByNetwork(network: EncryptPayloadNetwork, content: string | Uint8Array): string {
    // v38
    if (typeof content === 'string') {
        if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
        return content
    }

    // v37
    if (network === EncryptPayloadNetwork.Twitter) return __TwitterEncoder(content)
    return sharedEncoder(content)
}

export * from './twitter.js'
