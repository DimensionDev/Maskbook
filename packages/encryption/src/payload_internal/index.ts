import type { CheckedError } from '@masknet/shared-base'
import type { Result } from 'ts-results'
import type { PayloadParseResult } from '..'
import type { CryptoException, PayloadException } from '../types'

export type PayloadParserResult = Promise<
    Result<PayloadParseResult.Payload, CheckedError<CryptoException | PayloadException>>
>
export { parse40 } from './version-40'
export { parse39 } from './version-39'
export { parse38 } from './version-38.parser'
// Remove v38 encoder after
export { encode38 } from './version-38.encoder'
export { parse37 } from './version-37.parser'
export { encode37 } from './version-37.encoder'
