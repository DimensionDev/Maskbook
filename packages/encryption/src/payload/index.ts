import type { PayloadWellFormed } from '..'
import { encode37, encode38, parse37, parse38, parse39, parse40, PayloadParserResult } from '../payload_internal'
import { EKindsError as Err, EKinds, EKindsError } from '../types'

export * from './types'
export async function parsePayload(payload: unknown): PayloadParserResult {
    if (payload instanceof ArrayBuffer) {
        return parse37(payload)
    }
    if (typeof payload === 'string') {
        if (payload.startsWith('ðŸŽ¼4/4')) return parse38(payload)
        if (payload.startsWith('ðŸŽ¼3/4')) return parse39(payload)
        if (payload.startsWith('ðŸŽ¼2/4')) return parse40(payload)
    }
    return new Err(EKinds.InvalidPayload, 'Unknown version').toErr()
}

export function encodePayload(payload: PayloadWellFormed.Payload) {
    if (payload.version === -38) return encode38(payload)
    else if (payload.version === -37) return encode37(payload)
    return new EKindsError(EKinds.EncodeFailed, 'unsupported payload version').toErr()
}
