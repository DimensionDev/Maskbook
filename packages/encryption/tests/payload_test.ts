import { test, expect } from '@jest/globals'
import { parsePayload } from '../src'

test('1 + 1 == 2', async () => {
    expect((await parsePayload({})).err).toBeTruthy()
})
