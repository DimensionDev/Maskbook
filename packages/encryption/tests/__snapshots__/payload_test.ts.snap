// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple payload: Empty payload 1`] = `
Uint8Array [
  147,
  0,
  196,
  88,
  151,
  208,
  219,
  192,
  192,
  192,
  192,
  147,
  0,
  146,
  167,
  65,
  50,
  53,
  54,
  71,
  67,
  77,
  217,
  43,
  74,
  82,
  104,
  114,
  82,
  75,
  121,
  107,
  109,
  110,
  109,
  51,
  83,
  98,
  117,
  78,
  119,
  54,
  79,
  99,
  88,
  70,
  95,
  106,
  105,
  119,
  48,
  103,
  73,
  108,
  87,
  51,
  81,
  105,
  87,
  78,
  86,
  48,
  49,
  106,
  101,
  97,
  69,
  196,
  16,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  196,
  5,
  0,
  1,
  2,
  3,
  4,
  192,
]
`;

exports[`Simple payload: Empty payload parsed 1`] = `
Object {
  "author": Ok {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "authorPublicKey": Ok {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "encrypted": Ok {
    "err": false,
    "ok": true,
    "val": Uint8Array [
      0,
      1,
      2,
      3,
      4,
    ],
  },
  "encryption": Ok {
    "err": false,
    "ok": true,
    "val": Object {
      "AESKey": Ok {
        "err": false,
        "ok": true,
        "val": Object {
          "algr": "A256GCM",
          "key": CryptoKey {
            Symbol(kKeyObject): SecretKeyObject {
              Symbol(kKeyType): "secret",
            },
            Symbol(kAlgorithm): Object {
              "length": 256,
              "name": "AES-GCM",
            },
            Symbol(kExtractable): true,
            Symbol(kKeyUsages): Array [
              "encrypt",
              "decrypt",
            ],
          },
        },
      },
      "iv": Ok {
        "err": false,
        "ok": true,
        "val": Uint8Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
        ],
      },
      "type": "public",
    },
  },
  "signature": Ok {
    "err": false,
    "ok": true,
    "val": NoneImpl {
      "none": true,
      "some": false,
    },
  },
  "version": -37,
}
`;
